### DISCUSSION TITLE
[mini.hues] What is the recommended method to add additional highlight groups?

### DISCUSSION THREAD

**Discussion by pkazmier** [nvim-mini/mini.nvim#784] [Q&A] [answered]
What is the recommended method of adding additional highlight groups using the same palette on top of the existing ones provided by `mini.hues`? 

In most themes I've used before, usually there is a means to add your own highlights via a function that is called with the theme's palette of colors, so one can define/override groups using the same palette.

Would the current recommended approach be:

1. Call `setup` with standard config.
2. Call `make_palette` with the same config. Seems silly as `setup` already did this, but I have no means to obtain a reference to the palette.
3. Make a bunch of `vim.api.nvim_set_hl` calls using colors referenced from step 2.

One last question, the documentation has a note about how `mini.hues` doesn't create a color scheme. It then provides instructions on how to do so. I'm not sure I'm smart enough to understand the implications of this. Should I care? If I just call `setup` and it sets my "theme", do I need to create a real color scheme?

Thanks.

**Comment by echasnovski** [comment, marked as answer]
> What is the recommended method of adding additional highlight groups using the same palette on top of the existing ones provided by `mini.hues`?
> 
> In most themes I've used before, usually there is a means to add your own highlights via a function that is called with the theme's palette of colors, so one can define/override groups using the same palette.
> 
> Would the current recommended approach be:
> 
> 1. Call `setup` with standard config.
> 
> 2. Call `make_palette` with the same config. Seems silly as `setup` already did this, but I have no means to obtain a reference to the palette.
> 
> 3. Make a bunch of `vim.api.nvim_set_hl` calls using colors referenced from step 2.

Yes, this is the recommended way at the moment, if you specifically want to use palette colors.

Another, possibly more robust approach, would be to link to an existing group with similar semantics or visuals. Good examples will be 5 `DiagnosticXxx` groups representing five levels of attention and which all will have different text colors. So it will be something like `vim.api.nvim_set_hl(0, 'MyGroup', { link = 'DiagnosticHint' })`.

> One last question, the documentation has a note about how `mini.hues` doesn't create a color scheme. It then provides instructions on how to do so. I'm not sure I'm smart enough to understand the implications of this. Should I care? If I just call `setup` and it sets my "theme", do I need to create a real color scheme?

Usually there is no difference, but there might be. Creating separate color scheme has (at least) these major side effects:
- It triggers `ColorSchemePre` and `ColorScheme` events. Plugins might have autocommands for them to cleanup and adjust their highlight groups (usually if they are not linked to some already existing ones).
- It is meant to define `vim.g.colors_name` global variable indicating that color scheme was loaded. Some plugins might rely on this when setting their default highlight groups.
- It can be loaded with `:colorscheme` command.

**Comment by pkazmier** [comment]
> Yes, this is the recommended way _**at the moment**_, if you specifically want to use palette colors.

It would be nice if the palette could be exposed at some point.

> Another, possibly more robust approach, would be to link to an existing group with similar semantics or visuals. Good examples will be 5 DiagnosticXxx groups representing five levels of attention and which all will have different text colors. So it will be something like vim.api.nvim_set_hl(0, 'MyGroup', { link = 'DiagnosticHint' }).

Yes, good advice. Will be sure to try that approach first. During this process of trying to adopt mini, I've come across two cases so far where I need to define specific groups:

1. For the `MiniHipatterns*Colon` and `MiniHipatterns*Body` groups to accomplish https://github.com/echasnovski/mini.nvim/discussions/783
2. For the custom `mini.statusline` path section from https://github.com/echasnovski/mini.nvim/discussions/36#discussioncomment-8889147 as I now need a `MiniStatuslineFilenameModified` and `MiniStatuslineDirectory` groups.

**Comment by abeldekat** [comment]
> ... Call make_palette with the same config. Seems silly as setup already did this, but I have no means to obtain a reference to the palette.

A bit hacky: It's possible to obtain the reference using a closure in a decorated Hues.make_palette:

```lua
local Hues = require("mini.hues") -- "capture" the palette on hues.setup
local orig_make_palette = Hues.make_palette
local p = nil

---@diagnostic disable-next-line: duplicate-set-field
Hues.make_palette = function(config)
  p = orig_make_palette(config)
  return p
end

vim.api.nvim_create_autocmd("Colorscheme", {
  pattern = { "*hue" },
  callback = function()
    if p == nil then return end
    local hi = function(name, data) vim.api.nvim_set_hl(0, name, data) end
    -- Links to MiniJump2dSpot by default, changed fg(fg_edge2)
    hi("MiniJump2dSpotUnique", { fg = p.orange, bg = p.bg_edge2, bold = true, nocombine = true })
    -- Area for messages and cmdline, using Comment.fg
    hi("MsgArea", { fg = p.fg_mid2 })
  end,
})
```

I call this config before the `:colorscheme` command.

**Comment by pkazmier** [comment]
For any future reader,`get_palette` exists now: https://github.com/echasnovski/mini.nvim/commit/f4ebff007fdf9f94bccdf9904d50a91f6aec1bf8

**Reply by echasnovski** [reply]
Yeah, it is a bit awkward to not have `get_palette()` after the introduction of `apply_palette()`. So decided to add it.
